src/webgpu/api/validation/encoding/programmable/pipeline_bind_group_compat.spec.ts

```diff
@@ -793,3 +912,165 @@ g.test('empty_bind_group_layouts_requires_empty_bind_groups,render_pass')
       encoder.finish();
     }, !success);
   });
+
+g.test('default_bind_group_layouts_never_match,compute_pass')
+  .desc(
+    `
+  Test that bind groups created with default bind group layouts never match other layouts, including empty bind groups.
+
+  * Test that a pipeline with an explicit layout can not be used with a bindGroup from an auto layout
+  * Test that a pipeline with an auto layout can not be used with a bindGroup from an explicit layout
+  * Test that an auto layout from one pipeline can not be used with an auto layout from a different pipeline.
+
+  TODO:
+  * Test empty bindgroup layouts on the same default layout pipeline are not compatible. In other words if
+    you only define group(2) then group(0)'s empty layout and group(1)'s empty layout should be incompatible.
```

Comment:
The spec rules as written only require it to come from the same pipeline, not the same slot:
https://gpuweb.github.io/gpuweb/#bind-group-compatibility
This is a little odd, but probably fine as I don't think it's much of a footgun?

Reply:
In that case we need to test that 2 bind groups that appear the same, are the same? In other words

```
      @group(0) @binding(0) var s1: sampler;
      @group(0) @binding(1) var t1: texture_2d<f32>;

      @group(1) @binding(0) var s2: sampler;
      @group(1) @binding(1) var t2: texture_2d<f32>;
```

If we use `layout: 'auto'` on this then a bindGroup made with `pipeline.getBindGroupLayout(0)` should work bound to bind group 1.

I don't think there is a test for that and it would be easy for browsers to disagree on this without a test?

It does make me wonder why they aren't just compatible in general. The rules are spelled out for what gets generated.

Reply:
good point, we would definitely need to test that. Unless we want to make a breaking change in the spec and make it stricter.

The exclusivePipeline thing was added as a way to prevent footguns because auto layouts are sensitive to static uses in WGSL which are fairly subtle and easy to change (like by commenting out a function call, it removes from the BGL some binding that was used by that function's call tree). But I do wonder if it's really needed. Maybe with some good error messages there would be no footgun, like, "bindgroup created with auto BGL from pipeline 'X' but is used with pipeline 'Y' which has an incompatible auto BGL. Note that auto BGLs do not include declared bindings which are not reachable from the entrypoint of the pipeline".
Jasper and Brandon both had more opinions about this. We definitely picked the current behavior knowing it was a more "safe for 1.0" decision that we didn't necessarily have to keep forever.